### PR TITLE
Improve flexibility  of appModules discovery by allowing to map arbitrary appModule to a given executable.

### PR DESCRIPTION
**Pull request by lukaszgo1**
### Link to issue number:
Related to #13364

### Summary of the issue:
Currently when NVDA tries to load an appModule for a given program it just looks for a Python module of the same name as the program's executable. This has its limitations in particular:
- Some names are incompatible with the Python's import system (for example see #5323 where we resorted to replacing dots with underscores before attempting the import).
- When a single apModule should be used for a multiple executable's we need as many appModules as there are executable's (this also often causes linter failures as these alias modules just star imports everything from the main module).
- Even if the given module name does not contain symbols which are incompatible with `importlib` they may contain characters which are invalid in ordinary import statements (for one example see #13364 where the apModule is called "notepad++"). Since "+" is invalid in Python's import statement add-on developers cannot import this module from `nvdaBuiltin` which means that it cannot be extended in an add-on.
### Description of how this pull request fixes the issue:
This PR introduces a mapping of executable names to appModules which is consulted before the given module is imported. It also adds a convenience functions for add-on developers which can be used to register or unregister a specific module for a given program. All alias apModules currently present in the source are marked as deprecated and application's for which they were loaded are mapped to the right appModule using the new map.
Since we cannot remove the old alias app modules care has been taken not to use them - they' re kept only for add-ons developers.
### Testing strategy:
Made sure that right modules are loaded for Visual Studio Code Insiders and Media Player Classic Home Cinema x64 - both of these were using alias appModules previously.
### Known issues with pull request:
None known
### Change log entries:
For Developers
- A new convenience functions `registerExecutableWithAppModule` and `unregisterExecutable` were added to the `appModuleHandler` module. They can be used to use a single appModule with multiple executable's.
- Some alias appModules are marked as deprecated and would be removed in 2023.1. If your code imported from one of them (consult the list below) you should import from the replacement module.

| Removed module name | Replacement module |
| --- | --- |
| azuredatastudio | code |
| azuredatastudio-insiders | code |
| calculatorapp | calculator |
| code - insiders | code |
| commsapps | hxmail |
| dbeaver | eclipse |
| digitaleditionspreview | digitaleditions |
| esybraille | esysuite |
| hxoutlook | hxmail |
| miranda64 | miranda32 |
| mpc-hc | mplayerc |
| mpc-hc64 | mplayerc |
| notepad++ | notepadPlusPlus |
| searchapp | searchui |
| searchhost | searchui |
| springtoolsuite4 | eclipse |
| sts | eclipse |
| teamtalk3 | teamtalk4classic |
| textinputhost | windowsinternal_composableshell_experiences_textinput_inputapp |
| totalcmd64 | totalcmd |
| win32calc | calc |
| winmail | msimn |
| zend-eclipse-php | eclipse |
| zendstudio | eclipse |

### Code Review Checklist:

- [X] Pull Request description:
  - description is up to date
  - change log entries
- [X] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [X] API is compatible with existing add-ons.
- [X] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [X] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by lukaszgo1** [comment]
@seanbudd  While this has been submitted pretty late in the release cycle it's necessary to fix #13311 and I'd say it is pretty low risk. I would appreciate if this can be included in 2022.1.

**Comment by josephsl** [comment]
Hi, another use case: some app executable names conflict with built-in module names e.g. time.exe for Clock app in Windows 10 and later. Thanks for looking into this.

**Comment by josephsl** [comment]
Hi,

Does the custom app module mapping succeed when NVDA is restarted while focused on an app module requiring a custom mapping for the executable name? A good app to test this is with Microsoft Weather app where it contains periods in the middle of the executable name. Also, note that this change will require changes to global plugins, too which are loaded way later than app module for the focused control.

Thanks.

**Comment by lukaszgo1** [comment]
> Does the custom app module mapping succeed when NVDA is restarted while focused on an app module requiring a custom mapping for the executable name? 

Yes, it does.

> note that this change will require changes to global plugins, too which are loaded way later than app module for the focused control.

Not really. To test this I've commented out the mapping for Media Player x64 in `apModuleHandler`, created a simple global plugin which just registers a right binary with a right appModule and tested that the right module is loaded both when the application is started after NVDA,d and when NVDA restarts when focused in MPCHC.
Have you tested this and found any issues?

**Comment by josephsl** [comment]
Hi,

An issue that should be resolved from add-ons themselves: suppose an app module bundled with an add-on imports everything from an app module that was deleted with this PR. This will cause the add-on version of the app module to fail to load because Python could not import a nonexistent module. This affects Windows App Essentials add-on as my policy is to import from the app module with the executable name that's included in supported Windows 10/11 releases (take modern keyboard, for example where TextInputHost.exe is the executable name of the emoji panel included in Windows 10 Version 2004 and later which are releases supported by the add-on). I will be able to change this from the add-on side as soon as this PR is approved and merged provided that I can also resolve the resulting lint issue from add-on code.

Also, can you provide code examples for use by add-on writers for sake of completeness?

Thanks.

File: source/appModuleHandler.py
Comment (josephsl) [inline source/appModuleHandler.py:63]: Actually, you can either:

* Remove commsaps as it is no longer deployed to the public
* Assign commsapps to hxmail as commsapps derives from the empty hxoutlook, which in turn derives from hxmail

File: source/appModuleHandler.py
Comment (MarcoZehe) [inline source/appModuleHandler.py:60]:
```suggestion
	# The Insider version of Visual Studio Code should use the module for the stable version.

```

File: source/appModuleHandler.py
Comment (lukaszgo1) [reply source/appModuleHandler.py]: Thanks - I've done the latter.

**Comment by CyrilleB79** [comment]
Please see https://github.com/nvaccess/nvda/pull/13364#issuecomment-1046116349 regarding the possibility to import a module containing an unusable character in its name, e.g. '+' or '.'.

It's a bit ugly but this seems doable.

**Comment by seanbudd** [comment]
Has the new API been tested with an addon yet?

The App Modules [sections of the developer guide](https://www.nvaccess.org/files/nvda/documentation/developerGuide.html#toc13) will need updating and potentially other documentation.

Is it possible that addons rely on extending aliases? If so, we need to include a changelog item for that.

**Comment by josephsl** [comment]
Hi, yes I have tested this change with Windows App Essentials which relies on aliased app modules. Any add-on which includes app modules that derive from alias app modules from NVDA Core must now derive from base app modules (say, hxmail instead of hxoutlook), a change I think is the responsibility of the add-on authors (I have prepared a change in Windows App Essentials add-on that was tested and ready to go as soon as this PR is reviewed, approved, and merged). Note that it might be possible that the executable name for a derive module may change in the future, in which case add-on authors may need to either register the new executable name with an existing app module or import the base app module from Core. A simpler approach would have been to define a list of characters that should be replaced by underscores or an acceptable character (say, calling Notepad++.exe app module notepad__.py), but this PR does open a possibility where it becomes possible to load an app module with executable names that conflict with built-in Python module names such as Windows 10/11 Clock app (time.exe and time.py cannot be used because it conflicts with built-in time.py module from Python library). I understand that this is a late change, but given that people use Notepad++ for various tasks (including NVDA Core and add-on development), I advise reviewing this soon. Thanks.

**Comment by seanbudd** [comment]
>Note that it might be possible that the executable name for a derive module may change in the future, in which case add-on authors may need to either register the new executable name with an existing app module or import the base app module from Core. 

Isn't that a similar problem to existing aliases? If an executable name changes, we need to create a new alias. This is a backwards compatible action.

**Comment by josephsl** [comment]
Hi, perhaps – I guess real-life data will tell us the full story (after this PR is merged). Thanks.

**Comment by seanbudd** [comment]
is this an API breaking change? does it have to be?

**Comment by lukaszgo1** [comment]
>     * Make the rules of the import priorities more complex that is:
>       
>       1. Split the map of appModule aliases into two one for add-ons and the second one for the core modules.
>       2. First try to lookup the module in the map for add-ons.
>       3. If there is no alias try to import the module from `appModules` package.
>       4. Finally if import fails check for alias in the mapping for the core modules and import from there.

This is now implemented.

**Comment by lukaszgo1** [comment]
I would like to ask what are NV Access plans with regard to this work. 
As I see it before this can be merged:
- [ ] The developer guide needs to be updated to document the new ways of associating an appModule with the given binary (@seanbudd said that some additional documentation should also be updated - could you elaborate what else needs updating in your opinion?).


Then there are two options:
## Merge this in 2022.1 in that case:
- [ ] An appropriate change log entry should be added to document  alias appModules that were removed in this PR.


## Merge this in 2022.2 in that case:
- [ ] The removed alias appModules should be restored.
- [ ] - [ ] An issue should be created to make sure that these aliases are removed in 2023.1
- [ ] Since core appModules in the filesystem take precedence over the one in the map it would be nice to prefer importing from the mapped aliases rather than from disk just to use the new import mechanism for the entire year and not just after the alias appModules are removed.

File: source/appModuleHandler.py
Comment (seanbudd) [inline source/appModuleHandler.py:56]: This dictionary might be better placed in `appModules/__init__.py` - just to be closer to the core app modules.
For a developer adding a new core appModule, checking `appModules/__init__.py` is more intuitive.
It also separates the "data" from this code file, and slows the file length growth of `source/appModuleHandler.py`.

File: source/appModuleHandler.py
Comment (seanbudd) [inline source/appModuleHandler.py:108]: Fixed extra space

```suggestion
"""Maps names of the executables to the names of the appModule which should be loaded for the given program.
```

**Review by seanbudd — COMMENTED** [review]
> The developer guide needs to be updated to document the new ways of associating an appModule with the given binary (@seanbudd said that some additional documentation should also be updated - could you elaborate what else needs updating in your opinion?).

The developer guide is the main change.
After searching *.t2t and *.md files for "AppModule", it also might be worth adding anything relevant to `devDocs\technicalDesignOverview.md` as well, such as aliasing/adding core appmodules.

The approach on this PR looks good to me.
I believe this change is safe enough to be included late in the release cycle and can be included in 2022.1 if final approval makes the beta.

Note that "Link to issue number" needs updating. Are there any related issues for this PR? I would imagine that this API change has been suggested before.

**Comment by lukaszgo1** [comment]
@seanbudd  This is ready for review. I've moved the map out of `appModuleHandler` and updated both developer guide and technical design overview.I could not find any issue proposing this exact change (there is some related discussion in #9119 and #10248 but #10248 is not fixed by this PR).

**Comment by seanbudd** [comment]
Hi @lukaszgo1 ,

Sorry for the confusion, we've decided that it would be helpful for addon authors can test this new API early in a dev cycle.
This is in case the API needs further refinement.
We'll be looking to merge this early as possible in 2022.2 dev cycle.
I'm happy to fixup the PR with the 2022.2 merge case outline in this comment https://github.com/nvaccess/nvda/pull/13366#issuecomment-1049688735

**Comment by lukaszgo1** [comment]
@seanbudd  Thanks - I'll make it ready for merging in 2022.2 myself, though implementing "Since core appModules in the filesystem take precedence over the one in the map it would be nice to prefer importing from the mapped aliases rather than from disk just to use the new import mechanism for the entire year and not just after the alias appModules are removed." is going to be rather painful. Marking as a draft until it is ready for inclusion in 2022.2

**Comment by lukaszgo1** [comment]
@seanbudd Sorry for the delay. This should be ready for review and merging in 2022.2.

File: devDocs/developerGuide.t2t
Comment (seanbudd) [inline devDocs/developerGuide.t2t:212]: This fixes a typo and makes a the section linkable. 

```suggestion
++ Associating App Modules with an executable ++[AssociatingAppModule]
```

File: devDocs/developerGuide.t2t
Comment (seanbudd) [inline devDocs/developerGuide.t2t:200]: This fixes a typo and makes a link to the section. Testing the link renders correctly is worth doing.

```suggestion
If you want to use a single App Module for a multiple executables, or the name of the executable conflicts with the standard Python import rules read ["Associating App Modules with an executable" #AssociatingAppModule]
```

**Review by seanbudd — COMMENTED** [review]
Is there a way to log a warning that these alias modules are deprecated when importing them?

Perhaps we should open an issue to track the deprecation process for these, and keep the aliases. i.e. let's not announce deprecation until we have a deprecation plan and are confident in this new API.

File: devDocs/developerGuide.t2t
Comment (XLTechie) [inline devDocs/developerGuide.t2t:200]: Should remove the "a" before multiple.

Also "ap Module" should be "app Module". Actually, it should probably be "app module" or "App Module". I'm not sure it's a good idea to introduce yet a third casing for this phrase.

```suggestion
If you want to use a single app Module for multiple executables, or the name of the executable conflicts with the standard Python import rules see the section on mapping an app Module to the executable below.
```

File: devDocs/developerGuide.t2t
Comment (XLTechie) [inline devDocs/developerGuide.t2t:213]:
```suggestion
As explained above, sometimes the default way of associating an app Module with an application is not flexible enough. Examples include:
```

File: devDocs/developerGuide.t2t
Comment (XLTechie) [inline devDocs/developerGuide.t2t:214]:
```suggestion
- You want to use a single app Module for various binaries (perhaps both stable and preview versions of the application should have the same accessibility enhancements)
```

File: devDocs/developerGuide.t2t
Comment (XLTechie) [inline devDocs/developerGuide.t2t:215]: Trying to make this one read more clearly. Not sure if this does it, but here's an attempt.

```suggestion
- The executable file is named in a way which conflicts with the Python naming rules. I.e. for an application named "time", naming the app Module "time.py" would conflict with the built-in module from the standard library
```

File: devDocs/developerGuide.t2t
Comment (XLTechie) [inline devDocs/developerGuide.t2t:219]:
```suggestion
For example to map the app Module named "time_app_mod" to the "time" executable the plugin may be written as follows:
```

File: source/appModules/totalcmd64.py
Comment (XLTechie) [inline source/appModules/totalcmd64.py:1]:
```suggestion
# Use the app module for Total Commander 32 bit version.

```

File: source/appModules/textinputhost.py
Comment (XLTechie) [inline source/appModules/textinputhost.py:10]: Shouldn't there be two spaces before the comment here?
```suggestion
from .windowsinternal_composableshell_experiences_textinput_inputapp import *  # NOQA

```

**Review by XLTechie — CHANGES_REQUESTED** [review]

File: source/appModules/textinputhost.py
Comment (lukaszgo1) [reply source/appModules/textinputhost.py:10]: This is not introduced as part of this PR, and since we intent to remove this alias in the next compatibility breaking release I'm not sure we should be fixing anything in these files.

File: source/appModules/totalcmd64.py
Comment (lukaszgo1) [reply source/appModules/totalcmd64.py:1]: Same as above - this is not introduced in this PR and there is no point in modifying alias app Modules.

**Comment by lukaszgo1** [comment]
@seanbudd  This is ready for review.

> Is there a way to log a warning that these alias modules are deprecated when importing them?

Yes, I've implemented this in the latest commits.

> 
> Perhaps we should open an issue to track the deprecation process for these, and keep the aliases. i.e. let's not announce deprecation until we have a deprecation plan and are confident in this new API.



As far as I can tell the part about which we may not be confident enough is "Is the API proposed in this PR good enough for add-on authors?". Even if it turns out we need to change the way in which binaries are mapped to the modules this should not make a difference to what happens to the alias modules which are currently in core (they should be removed regardless of how exactly binaries are mapped to the App Modules.

**Comment by lukaszgo1** [comment]
@XLTechie  Many thanks for your linguistic corrections!

**Comment by josephsl** [comment]
Hi, as far as Windows App Essentials goes (which employs alias app modules), I have resolved compatibility issues in dev snapshots and plan to release needed changes to stable channel this summer. Thanks.

**Comment by XLTechie** [comment]
> @XLTechie Many thanks for your linguistic corrections!

You're welcome. :)

File: source/appModuleHandler.py
Comment (seanbudd) [inline source/appModuleHandler.py:51]:
```suggestion
_CORE_APP_MODULES_PATH: os.PathLike = appModules.__path__[0]
```

File: source/appModuleHandler.py
Comment (seanbudd) [inline source/appModuleHandler.py:292]:
```suggestion
	except Exception:
		# Broad except since we do not know
		# what exceptions may be thrown during import / construction of the app Module.
```

**Review by seanbudd — COMMENTED** [review]
Generally LGTM, some minor fixes suggested. I agree that the aliases should be removed, but we should create a new issue to track / discuss the removal plan.

As @XLTechie mentioned, there's several instances where a third casing "app Module" has been introduced. I suggest a case sensitive find and replace to "App Module".

File: source/appModuleHandler.py
Comment (lukaszgo1) [reply source/appModuleHandler.py]: I'm not sure I follow why this improves things. Usually the comment explaining why something needs to be done precedes the code which is explained i.e. docstring is at the top of the described method etc.

**Comment by lukaszgo1** [comment]
> Generally LGTM, some minor fixes suggested. 

I've applied them.

> I agree that the aliases should be removed, but we should create a new issue to track / discuss the removal plan.

Please see #13627

> 
> As @XLTechie mentioned, there's several instances where a third casing "app Module" has been introduced. I suggest a case sensitive find and replace to "App Module".



Fixed as well.

File: source/appModuleHandler.py
Comment (seanbudd) [reply source/appModuleHandler.py]: I think this is just a personal opinion sorry, couldn't find anything PEP related.
I've found commenting within the scope to be generally more pythonic.
In your example, the doctstring is under the function header,, in this case, it's within the scope of the exception.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**
